R/row_policy.R

```diff
@@ -0,0 +1,196 @@
+#' }
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("my_policy") %>%
+#'   rls_run()
+#' rls_policies(con)
+#' 
+#' # cleanup
+#' rls_drop_policies(con)
+#' dbDisconnect(con)
+row_policy <- function(.data, name) {
+  pipe_autoexec(toggle = rls_env$auto_pipe)
+  assert_is(name, "character")
+  assert_scalar(name)
+  .data <- as_row_policy(.data)
+  .data$name <- name
+  .data
+}
+
+#' Commands
+#' 
+#' @export
+#' @inheritParams grant
+#' @examplesIf has_postgres()
+#' library(DBI)
+#' library(RPostgres)
+#' con <- dbConnect(Postgres())
+#' if (!dbExistsTable(con, "passwd")) {
+#'    setup_example_table(con, "passwd")
+#' }
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("their_policy") %>%
+#'   commands(update)
+#' 
+#' # cleanup
+#' dbDisconnect(con)
+commands <- function(.data, ...) {
+  pipe_autoexec(toggle = rls_env$auto_pipe)
+  .data <- as_row_policy(.data)
+  .data$commands <- dot_names(...)
+  .data
+}
+
+#' Create rule for existing rows
+#' 
+#' @export
+#' @inheritParams grant
+#' @param using an expression to use to check against existing rows
+#' @param sql (character) sql syntax to use for existing rows
+#' @details Use either `using` or `sql`, not both
+#' @examplesIf has_postgres()
+#' library(DBI)
+#' library(RPostgres)
+#' con <- dbConnect(Postgres())
+#' if (!dbExistsTable(con, "passwd")) {
+#'    setup_example_table(con, "passwd")
+#' }
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("a_good_policy") %>%
+#'   commands(update) %>%
+#'   rows_existing(sql = 'current_user = "user_name"')
+#' 
+#' # cleanup
+#' dbDisconnect(con)
+rows_existing <- function(.data, using = NULL, sql = NULL) {
+  pipe_autoexec(toggle = rls_env$auto_pipe)
+  using_quo <- enquo(using)
+  stopifnot("Can not using and sql parameters together" =
+    xor(!rlang::quo_is_null(using_quo), !is_empty(sql)))
+  .data <- as_row_policy(.data)
+  if (rlang::is_null(sql)) {
+    .data$existing_rows <- translate_sql(!!using_quo, con = as_con(.data))
+  } else {
+    .data$existing_rows <- sql
+  }
+  .data
+}
+
+#' Create rule for new rows
+#' 
+#' @export
+#' @importFrom dbplyr translate_sql
+#' @inheritParams grant
+#' @param check an expression to use to check against addition of
+#' new rows or editing of existing rows
+#' @param sql (character) sql syntax to use for new rows
+#' @details Use either `check` or `sql`, not both
+#' @examplesIf has_postgres()
+#' library(DBI)
+#' library(RPostgres)
+#' con <- dbConnect(Postgres())
+#' if (!dbExistsTable(con, "passwd")) {
+#'    setup_example_table(con, "passwd")
+#' }
+#'
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("a_policy") %>%
+#'   commands(update) %>%
+#'   rows_existing(TRUE) %>%
+#'   rows_new(TRUE) %>%
+#'   to(jane)
+#'
+#' rls_tbl(con, "passwd") %>%
+#'   row_policy("that_policy") %>%
+#'   commands(update) %>%
+#'   rows_existing(sql = 'current_user = "user_name"') %>%
+#'   rows_new(home_phone == "098-765-4321") %>%
+#'   to(jane)
+#' 
+#' # cleanup
+#' dbDisconnect(con)
```

Comment:
Same here, would you mind adding an example where you would use `check`?

Reply:
Will do

Reply:
so i think it's best to drop NSE for `rows_existing` and `rows_new` (which I did) - so not adding egs as youd expect